src/utp_stream.cpp

```diff
@@ -210,6 +210,152 @@ struct packet
 	std::uint8_t buf[1];
 };
 
+// single thread packet allocation packet pool
+// can handle 3 common cases of packet allocation size
+struct packet_pool
+{
+	packet *acquire(std::uint16_t allocate)
+	{
+#if TORRENT_USE_ASSERTS
+		check_thread();
+#endif
+		packet *p{ nullptr };
+		std::uint16_t should_allocate{ allocate };
+
+		if (allocate <= utp_header_size)
+		{
+			p = m_syn_stash.pop();
+			should_allocate = utp_header_size;
+		}
+		else if (allocate <= mtu_floor_size)
+		{
+			p = m_mtu_floor_stash.pop();
+			should_allocate = mtu_floor_size;
+		}
+		else if (allocate <= mtu_ceiling_size)
+		{
+			p = m_mtu_ceiling_stash.pop();
+			should_allocate = mtu_ceiling_size;
+		}
+
+		if (p == nullptr)
+			p = static_cast<packet*>(malloc(sizeof(packet) + should_allocate));
```

Comment:
it's not an obvious win, in my mind, to make the ``sizeof(packet)`` implied in the request for memory. From the users point of view, it may look like a bug to request less memory than what is used